ContentType BDD implementation

miguelcleverti opened the pull request:
Added support for for content type behat tests.
Content type creation, deletion and verifications

miguelcleverti commented:
@joaoinacio @andrerom @bdunogier

bdunogier commented:
As discussed on skype:
- change the `ContentType` trait to a normal `ContentTypeContext`.
- move the ContentTypeManager code to the `ContentTypeContext`

andrerom commented:
+1 with proposed changes.

miguelcleverti commented:
Changed the CotentType and ContentTypeGroup into contexts.
As discussed with @bdunogier, still missing the content clean up after the tests, it will be done on a different PR. If everything is OK I will do the same of the other contexts.
The basic idea behind this is, as @bdunogier said, to use composition instead of 'scattered' inheritance.
I also removed some of the sentences related to test creating ContentTypeGroup with a given ID, since it was not really creating Groups with the ID just making an internal mapping, IMO these sentences were not really useful.
@andrerom @joaoinacio

bdunogier commented:
Please don't rebase the PR on every change. It makes it impossible to see what was changed last.

bdunogier commented on Context/Context.php:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return array Associative with 'login' and 'password'
+     */
+    protected function getCredentialsFor($roleIdentifier)
```
I wouldn't call it "get" credentials, since it actually creates credentials.

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,335 @@
+            $contentTypeGroup = $this->contentTypeGroupContext->ensureContentTypeGroupExists($groupIdentifier);
+            $contentTypeInGroups = $contentType->getContentTypeGroups();
+
+            // if the content type exist in the provided group the returns the content type
```
typo ? it feels like a couple words are missing.

bdunogier commented:
I don't think I would tie ContentType and ContentTypeGroup that way. I'd just make sure ContentTypes are created in the default group (content) unless specified otherwise. It would simply a lot of things, and would avoid adding operations that might hide the real source of issues.

bdunogier commented:
The repository/sudo part does not really convince me, gotta admit. A lot of repetition, and makes the code really more complicated.

Could we do this differently ? Why don't we login the admin user ?

miguelcleverti commented:
@bdunogier added the proposed fixes. Also added the DB clean up.

crevillo commented on Features/contenttype.feature:
```diff
@@ -0,0 +1,16 @@
+            | title      | ezstring | Title |
+        Then Content Type with identifier "Test1" exists in Group with identifier "Content"
+
+    Scenario: Test a content exits
```
```a content type```?

crevillo commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,241 @@
+        }
+
+        if (!$contentType && $throwIfNotFound) {
+            throw new \Exception("Could not load content type with identifier '${identifier}'");
```
what's the ```${``` for here?

crevillo commented on Features/contenttype.feature:
```diff
@@ -0,0 +1,16 @@
+Feature: Content Type group Creation
```
not full understand this... if the feature is about Content Type group shouldn't be in contenttype_group? or the Feature should be ```Content Type Creation```?

miguelcleverti commented:
@crevillo thank you for the feedback. Fixed those them! :+1:

andrerom commented on Context/Context.php:
```diff
@@ -0,0 +1,62 @@
+     */
+    public function cleanDB($event)
+    {
+        $this->installer->createConfiguration();
```
note: this is being deprecated over in https://github.com/ezsystems/ezpublish-kernel/pull/1556

andrerom commented on Context/Context.php:
```diff
@@ -0,0 +1,62 @@
+    {
+        $repository = $this->repository;
+        $userService = $repository->getUserService();
+        $repository->setCurrentUser($userService->loadUser(self::ADMIN_USER_ID));
```
You can optionally simplify these three lines to avoid the loading of admin user all the time by doing:
```php
$this->repository->setCurrentUser(new UserReference(self::ADMIN_USER_ID));
```

bdunogier commented:
I would not have taken care of DB reset in the same PR, but I guess that it would be quite a bit of work to separate them now, would it ?

bdunogier commented:
After re-reading the code, I'd really prefer if you separated the DB cleanup changes from the ContentType part.

bdunogier commented:
+1, but I'd like it to pass on Travis before we merge. Can we do that ?

miguelcleverti commented:
@bdunogier now the standard suite only includes tests that use the traits and `EzContext`, when all the trait are removed this suite will go in the mean while this will keep the tests from breaking. Also added a `@broken` tag to the test that is failing, from what I can tell it is some what related to cache. So is this one ready?

miguelcleverti commented:
ping @bdunogier @andrerom @joaoinacio could use some final review.

joaoinacio commented on Context/Object/ContentTypeGroup.php:
```diff
@@ -9,70 +9,81 @@
-        $identifier = $this->findNonExistingContentTypeGroupIdentifier();
-
-        $this->ensureContentTypeGroupWithIdAndIdentifierExists( $id, $identifier );
+        $this->repository = $repository;
```
` $this->repository = $repository;`
doesn't seem used, missing parent construtor?

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,244 @@
+ */
+class ContentType extends Context
+{
+    const DEFAULT_LANGUAGE = 'eng-GB';
```
Shouldn't this be part of the parent context ?

bdunogier commented on Context/Context.php:
```diff
@@ -0,0 +1,43 @@
+use eZ\Publish\Core\Repository\Values\User\UserReference;
+
+/**
+ * Context has all the needed methods and helpers that are used in contexts
```
A very ambitious description given what it does now :-)

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,244 @@
+    /**
+     * @Given (that) a Content Type does not exist with identifier :identifier
+     *
+     * Makes sure a content type wit 'identifier' does not exist.
```
"with"

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,244 @@
+    /**
+     * @Then Content Type (with identifier) :identifier exists
+     *
+     * Verifies that a content type with 'identifier' exists
```
$identifier

bdunogier commented:
About phpdoc, I'd use the variable instead of a quoted string (ex: `with 'identifier'` => `with $identifier). This particular one would also sound better as `a content type identified by $identifier`.

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,244 @@
+        }
+
+        if (!$contentType && $throwIfNotFound) {
+            throw new \Exception("Could not load content type with identifier '$identifier'");
```
You could just throw the exception again, it would avoid loosing the stacktrace.

bdunogier commented on Context/Object/ContentType.php:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /**
+     * Creates a content type with 'identifier' on content type group with identifier 'groupIdentifier' and with the
```
See general remark about quoted variables names, and words like "with".

bdunogier commented:
+1, but remember to squash the commits if it applies (remember `git commit --fixup`) ?

miguelcleverti commented:
@bdunogier fixed the issues mentioned and squashed the commits :smile:

bdunogier commented:
And it is merged. Thank you @miguelcleverti.
